### PR TITLE
Add option to fit and use self profile for HorneExtract

### PR DESCRIPTION
**Pull request by cshanahan1**
This PR adds the 'self_profile' option to HorneExtract. Closes JDAT-3222.

A summary of how this works (row/col language assuming a horizontal trace):
- The image is divided along the dispersion axis into equal (plus minus one to make integer sized bins) bins along the x axis.
- Columns with masked or NaN values are omitted (see comments below, proper masking should be addressed eventually)
- The median in each bin along the x axis is computed, producing a median profile in that bin
- These samples are interpolated between (using RectBivariateSpline, by default linearly)
- The Horne normalization process now happens for each column. The interpolator is evaluated at that wavelength column to produce the spatial profile used for weighting. A 1D spectrum object is returned.

I also expanded the docs and fixed a few things in there (i.e a mask is not strictly required for HorneExtract as implied before, but is optional and currently limited in its functionality) 

A few comments, concerns I have for reviewers:
1. This, like the default Gaussian option, does not handle masking properly. This should be addressed in a follow up PR but for now I kept the same limitation that if there are any masked values in a wavelength column, the 1D spectrum will not be returned for that wavelength.
2. Should we return the fit profile (a 2D interpolator object, or the interpolator evaluated on the image grid) as an attribute on the extract object? Right now I have it as a private attribute for testing purposes, but I'm not sure if this is something users would go looking for?
4. The Gaussian option allows you to provide 'bkg_model' to fit simultaneously with the Gaussian extraction kernel, creating a compound model. I did not implement this for the self_profile option because it was not straightforward since I'm no longer using models/fitters and couldn't create a compound model to fit. I also disagree with this option in general, as the documentation and existence of background steps in Specreduce implies that background should be removed beforehand, so I didn't want to spend too much time figuring this out in the event that others agree with my opinion on this. If not, I will figure out how to add this option in a follow up PR.

### CODE REVIEW

File: specreduce/extract.py [hunk 1]
```diff
@@ -267,9 +281,26 @@ class HorneExtract(SpecreduceOperation):
...
-    bkgrd_prof : `~astropy.modeling.Model`, optional
-        A model for the image's background flux.
-        [default: models.Polynomial1D(2)]
+    bkgrd_prof : `~astropy.modeling.Model` or None, optional
+        A model for the image's background flux. If ``spatial_profile`` is set
+        to ``self_profile``, then ``bkgrnd_prof`` must be set to None.
+        [default: models.Polynomial1D(2)].
+
+    spatial_profile : str, optional
+        The shape of the object profile. Choise of 'gaussian' to fit a uniform
```
Comment (kecnry) [inline specreduce/extract.py:290]:
```suggestion
        The shape of the object profile. Choice of 'gaussian' to fit a uniform
```

File: specreduce/extract.py [hunk 2]
```diff
@@ -448,6 +572,22 @@ def __call__(self, image=None, trace_object=None,
         bkgrd_prof : `~astropy.modeling.Model`, optional
             A model for the image's background flux.
 
+        spatial_profile : str, optional
+            The shape of the object profile. Choise of 'gaussian' to fit a uniform
```
Comment (kecnry) [inline specreduce/extract.py:576]:
```suggestion
            The shape of the object profile. Choice of 'gaussian' to fit a uniform
```

File: specreduce/extract.py [hunk 3]
```diff
@@ -478,10 +618,22 @@ def __call__(self, image=None, trace_object=None,
...
+                                      interp_degree_self_profile is not None
+                                      else self.interp_degree_self_profile)
         variance = variance if variance is not None else self.variance
         mask = mask if mask is not None else self.mask
         unit = unit if unit is not None else self.unit
 
+        # make sure a valid profile type was specified:
+        spatial_profile = spatial_profile.lower()
+        if spatial_profile not in ['gaussian', 'self_profile']:
+            raise ValueError("spatial_profile must be 'gaussian' or 'self_profile'")
```
Comment (kecnry) [inline specreduce/extract.py:635]: Would this help future-proof for when we might add more valid options?

```suggestion
        spatial_profile_choices = ('gaussian', 'self_profile')
        if spatial_profile not in spatial_profile_choices:
            raise ValueError(f"spatial_profile must be one of {', '.join(spatial_profile_choices)}")
```

File: specreduce/extract.py [hunk 4]
```diff
@@ -497,50 +649,79 @@ def __call__(self, image=None, trace_object=None,
...
+                mean_init_guess = trace_object.trace
+            else:
+                mean_init_guess = np.broadcast_to(
+                    img.shape[crossdisp_axis] // 2, img.shape[disp_axis]
+                )
+
+        else:  # self_profile
+            # for now, bkgrnd_prof must be None because a compound model can't
+            # be created with a interpolator + model. i think there is a way
+            # around this, but will follow up later
```
Comment (kecnry) [inline specreduce/extract.py:683]: should we create a follow-up issue for this or is it already being tracked somewhere?

File: specreduce/extract.py [hunk 5]
```diff
@@ -497,50 +649,79 @@ def __call__(self, image=None, trace_object=None,
...
+            if bkgrd_prof is not None:
+                raise ValueError('When `spatial_profile`is `self_profile`,'
+                                 '`bkgrnd_prof` must be None. Background should'
+                                 ' be fit and subtracted from `img` beforehand.')
+            # make sure n_bins doesnt exceed the number of (for now) finite
+            # columns. update this when masking is fixed.
+            n_finite_cols = np.logical_or.reduce(or_mask, axis=crossdisp_axis)
+            n_finite_cols = np.count_nonzero(n_finite_cols.astype(int) == 0)
+
+            # determine interpolation degree from input and make
```
Comment (kecnry) [inline specreduce/extract.py:693]: and make what?

File: specreduce/extract.py [hunk 6]
```diff
@@ -497,50 +649,79 @@ def __call__(self, image=None, trace_object=None,
...
+            if isinstance(interp_degree_self_profile, int):
+                kx = ky = interp_degree_self_profile
+            else:  # if input is tuple of ints
+                kx, ky = interp_degree_self_profile
```
Comment (kecnry) [inline specreduce/extract.py:697]: does there need to be any type-checking (or length checking) here on `interp_degree_self_profile` to avoid cryptic errors?

File: specreduce/extract.py [hunk 7]
```diff
@@ -497,50 +649,79 @@ def __call__(self, image=None, trace_object=None,
...
+
+            if n_bins_self_profile >= n_finite_cols:
+
+                raise ValueError(f'`n_bins_self_profile` ({n_bins_self_profile}) '
+                                 'should not exceed the number of fully-finite '
```
Comment (kecnry) [inline specreduce/extract.py:702]: This wording could be confusing for the user when `n_bins_self_profile == n_finite_cols`

```suggestion
                                 'must be less than the number of fully-finite '
```

File: specreduce/extract.py [hunk 8]
```diff
@@ -549,18 +730,32 @@ def __call__(self, image=None, trace_object=None,
...
-            # fit compound model to column
-            fitted_col = fit_ext_kernel(xd_pixels)
-
-            # save result and normalization
-            kernel_vals.append(fitted_col)
-            norms.append(fit_ext_kernel.amplitude_0
-                         * fit_ext_kernel.stddev_0 * np.sqrt(2*np.pi))
+            if self.spatial_profile == 'gaussian':
+
+                # else, set compound model's mean to column's matching trace value
```
Comment (kecnry) [inline specreduce/extract.py:735]:
```suggestion
                # set compound model's mean to column's matching trace value
```

File: specreduce/extract.py [hunk 9]
```diff
@@ -549,18 +730,32 @@ def __call__(self, image=None, trace_object=None,
...
+                if bkgrd_prof is not None:  # attr names will be diff. if not compound
+                    fit_ext_kernel.mean_0 = mean_init_guess[col_pix]
+                else:
+                    fit_ext_kernel.mean = mean_init_guess[col_pix]
+
+                # evaluate fit model (with shifted mean, based on trace)
+                fitted_col = fit_ext_kernel(xd_pixels)
+
+                # save result and normalization
+                # this doesn't need to be in this loop, address later
```
Comment (kecnry) [inline specreduce/extract.py:746]: is there an issue tracking this?  If not, should we create one so we don't forget?

File: specreduce/tests/test_extract.py [hunk 10]
```diff
@@ -197,3 +239,91 @@ def test_horne_non_flat_trace():
...
+    # option add a perfectly gaussian source and make sure gaussian extraction
+    # and self-profile extraction agree (since self=gaussian in this case)
+
+    image = mk_test_img
+    add_gaussian_source(image)  # add source across image, flat trace
+
+    trace = FlatTrace(image, image.shape[0] // 2)
+
+    print(image.data)
+
```
Comment (kecnry) [inline specreduce/tests/test_extract.py:268]:
```suggestion
```

**Review by PatrickOgle — APPROVED** [review]
Performed science review on simulated 2D spectrum with tilted trace and expanding PSF.  This PR gives the proper flux level and improved SNR compared with the existing HorneExtract. SNR improves with profile binning as expected.

**Comment by PatrickOgle** [comment]
Would be useful to expose cross-dispersion profile as an output. (For future PR if not now.)

**Comment by cshanahan1** [comment]
working on rebasing this (which has become frustrating :| ) and addressing kyle's comments, will push changes shortly

**Comment by cshanahan1** [comment]
> Would be useful to give cross-dispersion profile as an output. (For future PR if not now.)

i have it as a private attribute (self._interp_spatial_prof) which i was using for testing, i can just remove the underscore and add a line to the docs about it if thats what you mean?

File: specreduce/extract.py (hunk 4 above)
Comment (cshanahan1) [reply specreduce/extract.py:683]: there's no follow up issue - i think we should consider removing this option altogether which is why i didn't spend the time to do this here.

File: specreduce/extract.py (hunk 5 above)
Comment (cshanahan1) [reply specreduce/extract.py:694]: lol, its a mystery (just kidding, fixed)

File: specreduce/extract.py (hunk 6 above)
Comment (cshanahan1) [reply specreduce/extract.py:698]: added to be safe

File: specreduce/extract.py (hunk 9 above)
Comment (cshanahan1) [reply specreduce/extract.py]: no, i plan to address this when i address masking. right now it makes sense for it to be in the loop, even though the area under the gaussian doesn't change when the mean shifts (and, we don't need to shift the mean at all since align_along_trace is accounting for this), but it does append NaN to the output array if the column has any masked values which is how masks are treated for now.

**Comment by eteq** [comment]
A few high-level API thoughts:

First, I think the use of keyword arguments that apply only for certain values of another keyword (i.e., `spatial_profile`) is a bit of an anti-pattern.  I see two immediate alternatives, although there are probably other options as well.
1. have `spatial_profile` accept an *object* and then gaussian and self-profile could be classes that take the various keywords as arguments.  This is closer to the idiom used in, say, `photutils` and `specutils` (at least in some places)
2. If that seems a bit too heavy-weight, another option is to use tuples or dicts in the same way.  E.g., `spatial_profile` could take a dict that is something like ``{'name': 'self_profile', 'n_bins_self_profile':5, ...}`` or similar.  That's a bit harder to document and requires some annoying validation logic, but is lighter-weight than a class.

Second, I find the name "self_profile" to be not great because "self" can mean a lot of things. My first thought was "empirical_binned_profile", but that's a bit wordy... maybe someone else has something a bit more compact?

Third, re:
> Should we return the fit profile
from @cshanahan1 - yes, I think so!  I think the user will indeed sometimes want this.

Last, re: background model:  I think the background model is very important for some use cases, but I see the complexity.  FWIW, this is yet another example of the problem that astropy modeling really needs an implementation of splines... but I think it's ok to make an issue that basically says that's the core of the problem, and then whenever that's possible, we'll add background support here, but otherwise not prioritizing it. Realistically, I think the empirical nature of this fit means the cases where background subtraction is important (strong skylines) is better suited to the gaussian case anyway.

**Comment by kecnry** [comment]
@eteq - if we want different input arguments, different exposed methods, and different defaults for background behavior.... it's feeling more and more to me like we should just have them as separate classes (likely with shared code, consistent APIs, and not necessarily needing to be split in this PR).

**Comment by cshanahan1** [comment]
Thanks for the feedback @eteq! What do you think about swapping to 'empirical_profile', I think that is a bit more informative.

Though its outside of the scope of this current ticket, there will eventually be at least a third independent option for spatial profile (that also won't share the 2 new keywords here, and may have its own). Keeping that in mind, it seems there are a few options for how to organize this

1. (What is currently in this PR) a single HorneExtract class, with an argument ('spatial_profile') that dictates which type will be used ('gaussian' and 'self_profile' for now, perhaps a 'user_supplied' option later). For each type, there are a set of keyword arguments that will not be used for other types (i.e. n_bins_self_profile is ignored if spatial_profile is 'gaussian', but does not throw an error if set). 

2. Also just one HorneExtract class, but 'spatial_profile' accepts a dictionary if there are associated keyword args or just a string if not. For example, since gaussian doesn't require any additional info you would set it by doing spatial_profile='gaussian', but to use the self profile option you could either do spatial_profile='self_profile' to use the default for n_bins and interp_degree, or spatial_profile={'method':'self_profile', 'n_bins_self_profile':50}. This has the benefit of only having one arg that controls the spatial profile, but would be more annoying to just quickly change the number of bins from the default since you'd have to change your input from a string to a dict and remember what keys it wants etc.

3. A single HorneExtract class, but with different 'call' methods for each extraction type. Constructing the Horne class would not take any of the spatial_profile arguments when created, those would be supplied to each HorneExtract().gaussian_profile_extraction, HorneExtract().empirical_profile_extraction()

4. Multiple HorneExtract classes, for each profile type.

Any other ideas? Or votes? I am leaning towards option 2.

**Comment by eteq** [comment]
`empirical_profile` is definitely better, although I think it's still not ideal since there's still several "empirical" things. But I don't have an obvious better suggestion...

Also, in a side-conversation, @kecnry @tepickering @rosteen and @camipacifici discussed the interface question and I think we settled on option 2 *for now* to finish this PR, but that we need to come to consensus on solving the problem here in a consistent way across specreduce, and we can re-jigger this to fit that once that decision is made.

**Comment by tepickering** [comment]
agreed, option 2 _for now_ is fine. ultimately, both `spatial_profile` and `background_profile` should have the same signature: `callable(wavelength) -> 1D array | callable(spatial)`. i think it's fine for `HorneExtract` to support simple defaults if those aren't specified, but real world data can be highly non-rectilinear and the design will need to handle that eventually.

File: CHANGES.rst [hunk 11]
```diff
@@ -4,6 +4,13 @@
 New Features
 ^^^^^^^^^^^^
 
+- Added 'self_profile' option for HorneExtract. If The ``self_profile`` option
```
Comment (tepickering) [inline CHANGES.rst:7]: i think `interpolated_profile` is a better name and more accurately descriptive of it.

File: specreduce/extract.py [hunk 12]
```diff
@@ -250,6 +251,19 @@ class HorneExtract(SpecreduceOperation):
...
     spectrum.
 
+    There are two options for fitting the spatial profile used for
+    extraction - by default, a 1D gaussian is fit and as a uniform profile
+    across the spectrum. Alternativley, the ``self profile`` option may be
+    chosen - when this option is chosen, the spatial profile will be sampled
+    at various locations (set by <>) and interpolated between to produce a
+    smoothly varying spatial profile across the spectrum.
+
+    If using the Gaussian optoin for the spatial profile, a background profile
```
Comment (tepickering) [inline specreduce/extract.py:261]: optoin -> option

File: specreduce/extract.py [hunk 13]
```diff
@@ -294,6 +325,9 @@ class HorneExtract(SpecreduceOperation):
     image: NDData
     trace_object: Trace
     bkgrd_prof: Model = field(default=models.Polynomial1D(2))
```
Comment (tepickering) [inline specreduce/extract.py:327]: i would rename this `background_profile` for consistency with `spatial_profile`.

File: specreduce/tests/test_extract.py [hunk 14]
```diff
@@ -14,17 +15,49 @@
...
+@pytest.fixture
+def mk_test_img(nrows=30, ncols=10):
+    image = np.ones(shape=(nrows, ncols))
+    for j in range(image.shape[0]):
+        image[j, ::] *= j
+    image = CCDData(image, unit=u.Jy)
+    return image
+
+
+def add_gaussian_source(image, amps=2, stddevs=2, means=None):
```
Comment (tepickering) [inline specreduce/tests/test_extract.py:29]: this reminds me that i need to refactor `specreduce.synth_data.make_2d_trace_image` to have `profile` as an input parameter. right now it's hard-coded to `Moffat1D`.

**Review by tepickering — CHANGES_REQUESTED** [review]
i do have a fundamental concern that building an interpolated spatial profile as a function of wavelength is outside the scope of a class called `HorneExtract`. the implementation here also implicitly assumes rectilinear data which is rarely the case in the real world. 

that said, this is a good first draft of the functionality and something that can be built on. i would prefer the use of `interpolated_profile` vs `self_profile`. `empirical_profile` as @eteq suggested also works. (ultimately, it should just be `spatial_profile` and should be treated the same internally regardless of model.) i would also address the lines that `codecov` catches as not covered by tests.

File: CHANGES.rst (hunk 11 above)
Comment (cshanahan1) [reply CHANGES.rst:7]: done

File: specreduce/extract.py (hunk 13 above)
Comment (cshanahan1) [reply specreduce/extract.py:326]: should we do this in another PR so it has its own change log entry for clarity since this changes the interface?

Comment (tepickering) [reply specreduce/extract.py:326]: yeah, that's appropriate.

**Comment by cshanahan1** [comment]
@tepickering @kecnry @PatrickOgle @eteq any objections to merging? i think i addressed all comments and i also just added some more test coverage

**Review by tepickering — APPROVED** [review]

**Comment by tepickering** [comment]
thanks for all this work! changes look good to me and the checks all pass now so merge away.
